rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,155 @@
+# RFC: Use `as` instead of `children` for slot render function
+
+Contributors: behowell
+
+## Summary
+
+This proposes modifying the slot render function API to pass a render function via the slot's `as` prop instead of the `children` prop. This has several benefits, the primary of which is that it allows the user to write a render function for a slot without overriding the default children of a slot.
+
+This is a complementary proposal to bsunderhus's [RFC #27164: slot children render function support](https://github.com/microsoft/fluentui/pull/27164). It provides an alternative way to avoid having a render function override a slot's default `children` prop. This RFC has no dependencies on it, and could be adopted separately if needed.
+
+## Background
+
+Slots accept render functions that allow them to override the component and/or props used when rendering the slot. For more background, see https://react.fluentui.dev/?path=/docs/concepts-developer-customizing-components-with-slots--page#replacing-the-entire-slot
+
+The render function is passed in as the slot's `children` prop.
+
+## Problem statement
+
+A few problems with render functions as they are today:
+
+1. Using `children` as the render function will override any default children provided by the component's internals.
+2. It is confusing that passing a render function as `children` would override the _whole slot_, and not just its children.
```

Comment:
_(I'm assuming you meant to spread props in the render function `<div {...props}>Bar</div>`. It doesn't change the result, but it changes how I reply to the comment :)_

The render function chooses whether to override (or not) any props it's given. In this case, it's overriding children by specifying its own children for the div. So the result would be `Bar` rendered, and not `Foo`.

FWIW, I'm not necessarily _recommending_ that you use the slot's `children` prop rather than just putting children inside the render function. However, it is _possible_ with this RFC, whereas before it was not possible.

I.e. these two will produce exactly the same result:
```jsx
<>
  <Button icon={{ as: (_, props) => <div {...props}>Foo</div> }} />
  <Button icon={{ as: (_, props) => <div {...props} />, children: 'Foo' }} />
</>
```

Reply:
The same thing would happen if you provided any other prop to the slot, and then overrode it in the render function. This RFC just makes `children` work like any other prop:
```jsx
<Button icon={{ as: (_, props) => <div {...props} role="none" />, role: 'img' }} />
```

Reply:
Correct, so it will introduce two ways of doing things. With `children` you do not have the same issue:

```tsx
function App() {
  return (
    <>
      {/* Use default children */}
      <Button icon={{ children: (_, props) => <div {...props} /> }} />
      {/* Use override */}
      <Button icon={{ children: (_, props) => <div {...props}>Foo</div> }} />
    </>
  );
}
```

Reply:
It would make `children` like any other prop. Every prop has two ways of doing it: you could add it in slot props, or you could add it in the render function, like my example with the `role` prop.

Reply:
To clarify, having multiple ways to set `children` is _not a goal_ of this RFC. But given that all other props work the same way, I don't think it's a reason to avoid this change.